pkg/openstack/instancesv2.go

```diff
@@ -0,0 +1,226 @@
+*/
+
+package openstack
+
+import (
+	"context"
+	"fmt"
+	sysos "os"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	v1 "k8s.io/api/core/v1"
+	expirationcache "k8s.io/client-go/tools/cache"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/cloud-provider-openstack/pkg/client"
+	"k8s.io/cloud-provider-openstack/pkg/metrics"
+	"k8s.io/cloud-provider-openstack/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+const cacheTTL = 60 * time.Minute
+
+var cache expirationcache.Store
+
+func init() {
+	cache = expirationcache.NewTTLStore(stringKeyFunc, cacheTTL)
+}
+
+// InstancesV2 encapsulates an implementation of InstancesV2 for OpenStack.
+type InstancesV2 struct {
+	compute          *gophercloud.ServiceClient
+	region           string
+	regionProviderID bool
+	networkingOpts   NetworkingOpts
+}
+
+// InstancesV2 returns an implementation of InstancesV2 for OpenStack.
+func (os *OpenStack) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	if !os.useV1Instances {
+		return os.instancesv2()
+	}
+	return nil, false
+}
+
+func (os *OpenStack) instancesv2() (*InstancesV2, bool) {
+	klog.V(4).Info("openstack.Instancesv2() called")
+
+	compute, err := client.NewComputeV2(os.provider, os.epOpts)
+	if err != nil {
+		klog.Errorf("unable to access compute v2 API : %v", err)
+		return nil, false
+	}
+
+	regionalProviderID := false
+	if isRegionalProviderID := sysos.Getenv(RegionalProviderIDEnv); isRegionalProviderID == "true" {
+		regionalProviderID = true
+	}
+
+	return &InstancesV2{
+		compute:          compute,
+		region:           os.epOpts.Region,
+		regionProviderID: regionalProviderID,
+		networkingOpts:   os.networkingOpts,
+	}, true
+}
+
+// InstanceExists indicates whether a given node exists according to the cloud provider
+func (i *InstancesV2) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	_, err := i.getInstance(ctx, node)
+	if err == cloudprovider.InstanceNotFound {
+		klog.V(6).Infof("instance not found for node: %s", node.Name)
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+func (i *InstancesV2) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	server, err := i.getInstance(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	// SHUTOFF is the only state where we can detach volumes immediately
+	if server.Status == instanceShutoff {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+type MetadataCache struct {
+	Name     string
+	Metadata *cloudprovider.InstanceMetadata
+}
+
+// InstanceMetadata returns the instance's metadata.
+func (i *InstancesV2) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	obj, exists, err := cache.GetByKey(node.Name)
+	if err != nil {
+		klog.Warningf("InstanceMetadata cache lookup failure: %v", err)
+	}
```

Comment:
As mentioned on slack, could it be simpler instead to call this less frequently on large installations by tuning `NodeStatusUpdateFrequency`? I believe it would result in the same reduction in OpenStack api calls, but it would use the existing tuning options and it wouldn't be hard-coded to 60 minutes.

Reply:
@mdbooth thanks for the tip, that might work. With 60m cache it did (8 nodes) 288 requests in 12 hours towards openstack api, which means 24 requests hour -> 24/8 = 3 request/hour. 

I am now running ccm with `--node-status-update-frequency=60m` (and cache deleted).

However, the only problem is that how to communicate that people should really think of increasing the node status update frequency, because the default does not make sense in all situations. Like we are using kOps to provision our clusters and it means that its impossible to update addresses / flavor size in existing virtual machine (or impossible with kops, of course its possible to modify those manually in openstack.. but that is now how it should be used with kOps).